### PR TITLE
CR-423 Added utility class to read yml files

### PR DESCRIPTION
**Pull request by PeterBochel**
Added class to read yml file and replace node values with the contents of environment variables.

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        traverseAndUpdateProperties(childNode, createPathToNode(pathToNode, key));
+      } else {
+        // Found a property
+        attemptNodeUpdate(node, pathToNode, key);
+      }
+    }
+    ;
+  }
+
+  private void attemptNodeUpdate(JsonNode parentNode, String pathToParent, String nodeName)
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: attempt?

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  /**
+   * This method uses Jackson to convert the property data into a Java object.
+   *
+   * @param clazz is the class to convert the data into.
+   * @return Object containing the configuration date.
+   * @throws CTPException if the conversion failed.
+   */
+  public <T> T convertToObject(Class<T> clazz) throws CTPException {
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: being picky but public methods at top of class between constructor and privates

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ * 1) convert to lower case 2) replace '.', '-' or '_' with '#'
+ *
+ * <p>This means that the value of the field 'rabbitmq: username' will be replaced by the contents
+ * of the environment variable named 'RABBITMQ_USERNAME'. It can also be replaced with the contents
+ * of 'RabbitMQ.UserName'.
+ */
+public class OverrideableYmlConfigReader {
+  private static final Logger log = LoggerFactory.getLogger(OverrideableYmlConfigReader.class);
+
+  Map<String, String> envVariables = null;
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: private

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    // Use Jackson to read in the content of the yml file
+    JsonNode properties = readYmlFile(resourcePath);
+
+    // Replace property values with those from environment variables
+    String rootNodeName = "";
+    traverseAndUpdateProperties(properties, rootNodeName);
+    this.updatedProperties = properties;
+  }
+
+  private Map<String, String> getCleanedEnvironmentVariables() throws CTPException {
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: normaliseEnvironmentVariables

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+ * environment variable.
+ *
+ * <p>To match the names of environment variables to yml node names the following rules are applied:
+ * 1) convert to lower case 2) replace '.', '-' or '_' with '#'
+ *
+ * <p>This means that the value of the field 'rabbitmq: username' will be replaced by the contents
+ * of the environment variable named 'RABBITMQ_USERNAME'. It can also be replaced with the contents
+ * of 'RabbitMQ.UserName'.
+ */
+public class OverrideableYmlConfigReader {
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: YmlConfigReader

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java (hunk 1 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: done

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java (hunk 3 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: done

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java (hunk 4 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: done

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java (hunk 5 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: done

File: src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java (hunk 2 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java]: done

**Pull request merged**